Add `platform.getSubnet` API

felipemadero opened the pull request:
## Why this should be merged
There is no API related way to get subnet information such as transform subnet tx id, or updated owners

## How this works
It adds a new API call and client method, GetSubnet

## How this was tested
Tested on CLI branch developed to change subnet owners.

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -511,6 +511,83 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+
+// GetSubnetResponse is the response from calling GetSubnet
+type GetSubnetResponse struct {
+	// ID of the subnet
+	SubnetID ids.ID `json:"subnetID"`
```
Why do we return the subnetID in the response? Won't the caller already know it from the request?

felipemadero replied:
Yup. I had some confusion when making the rpc call with incorrect field name in the json (my mistake). The json was validated and the call returned info for platform without notice on my side. So I added the subnet ID. But it is best to just remove this and I also like returning error when requesting for platform (it is a way to validate the json param is correct).

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -511,6 +511,83 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+	SubnetID ids.ID `json:"subnetID"`
+	// whether it is permissioned or not
+	IsPermissioned bool `json:"isPermissioned"`
+	// subnet auth information for a permissioned subnet
+	ControlKeys []string    `json:"controlKeys"`
+	Threshold   json.Uint32 `json:"threshold"`
```
I think we would need to include the locktime here as well

felipemadero replied:
what is it used for?

StephenButtolph replied:
https://github.com/ava-labs/avalanchego/blob/master/vms/secp256k1fx/fx.go#L183-L184

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -511,6 +511,83 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+
+	response.SubnetID = args.SubnetID
+
+	if response.SubnetID == constants.PrimaryNetworkID {
+		response.IsPermissioned = false
+		response.ControlKeys = []string{}
+		response.Threshold = json.Uint32(0)
+		response.SubnetTransformationTxID = ids.Empty
+		return nil
+	}
```
I feel like we should just return an error here. The primary network isn't a subnet... (also these checks don't require the lock to be held)

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -511,6 +511,83 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+		return nil
+	}
+
+	if subnetTransformationTx, err := s.vm.state.GetSubnetTransformation(response.SubnetID); err == nil {
```
We should check for the actual error type - so that we don't report an unexpected error as incorrect information.

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -511,6 +511,83 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+
+	if subnetTransformationTx, err := s.vm.state.GetSubnetTransformation(response.SubnetID); err == nil {
+		response.IsPermissioned = false
+		response.ControlKeys = []string{}
+		response.Threshold = json.Uint32(0)
```
We should still populate these fields for permissionless subnets

danlaine commented:
Can we make a corresponding PR to update the docs with this new method?

felipemadero commented:
> Can we make a corresponding PR to update the docs with this new method?

here it is:

https://github.com/ava-labs/avalanche-docs/pull/1658

felipemadero commented:
> > Can we make a corresponding PR to update the docs with this new method?
> 
> here it is:
> 
> [ava-labs/avalanche-docs#1658](https://github.com/ava-labs/avalanche-docs/pull/1658)

@StephenButtolph @danlaine addressed your comments

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -510,6 +510,80 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+	subnetID := args.SubnetID
+
+	if subnetID == constants.PrimaryNetworkID {
+		return errors.New("the primary network isn't a subnet")
```
We should define this error with the rest of the errors in this file.

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -510,6 +510,80 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+}
+
+func (s *Service) GetSubnet(_ *http.Request, args *GetSubnetArgs, response *GetSubnetResponse) error {
+	s.vm.ctx.Log.Debug("API called",
+		zap.String("service", "platform"),
+		zap.String("method", "getSubnet"),
+	)
```
Can we add `zap.Stringer("subnetID", args.SubnetID)` here?

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -510,6 +510,80 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+		zap.String("method", "getSubnet"),
+	)
+
+	subnetID := args.SubnetID
+
```
stylistically I probably would just use `args.SubnetID` in the two places below... but 🤷

StephenButtolph commented on vms/platformvm/service.go:
```diff
@@ -510,6 +510,80 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
 	return nil
 }
 
+/*
+ *****************************************************
+ ******************* Get Subnet **********************
+ *****************************************************
+ */
+
```
I don't feel like these comments add anything anymore... (and they are really only around some of the old functions)

```suggestion
```

StephenButtolph commented:
just nits then lgtm

abi87 commented on vms/platformvm/service.go:
```diff
@@ -510,6 +511,73 @@ func (s *Service) GetUTXOs(_ *http.Request, args *api.GetUTXOsArgs, response *ap
+	Threshold   json.Uint32 `json:"threshold"`
+	Locktime    json.Uint64 `json:"locktime"`
+	// subnet transformation tx ID for a permissionless subnet
+	SubnetTransformationTxID ids.ID `json:"subnetTransformationTxID"`
```
q: I wonder if we should return the whole parameters here instead of simply the txID? In such a casa I would return also Primary network config when primary network ID is specified.
I am not sure we expose this config via API currently and it'd be a more uniform handling of subnets and primary network

felipemadero replied:
probably it is best to return formatted params instead of requiring the user to parse the tx. if @StephenButtolph agrees will make the change. also, regarding primary network params.

StephenButtolph replied:
I'd be willing to mrege this PR as is. If we wanted to return more information about permissionless subnets I wouldn't necessarily object.
